src/SourceBuild/tarball/content/patches/roslyn-analyzers/0002-Fix-ambiguous-DocumentBasedFixAllProvider-reference.patch

```diff
@@ -0,0 +1,47 @@
+From 8495f3ff6f49a1d923ee24d0d0e625783220ba6d Mon Sep 17 00:00:00 2001
+From: Michael Simons <msimons@microsoft.com>
+Date: Mon, 26 Jul 2021 16:33:22 +0000
+Subject: [PATCH] Fix ambiguous DocumentBasedFixAllProvider reference
+
+This addresses ambiguous reference issues related to DocumentBasedFixAllProvider.
+
+Example error:
+src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/Fixers/PreferIsKindFix.cs(51,53): error CS0104: 'DocumentBasedFixAllProvider' is an ambiguous reference between 'Analyzer.Utilities.DocumentBasedFixAllProvider' and 'Microsoft.CodeAnalysis.CodeFixes.DocumentBasedFixAllProvider'
+
+This is a result of https://github.com/dotnet/roslyn/commit/98e8cb016993a9bf3896e24f7d02947f6046ef97#diff-dd947268a990618d595914e376b02f059f6c761f61ce77eaa2bbf270f41ebbfc which publically exposed DocumentBasedFixAllProvider in Roslyn.
+
+'Microsoft.CodeAnalysis.CodeFixes.DocumentBasedFixAllProvider' is not compatible with `Analyzer.Utilities.DocumentBasedFixAllProvider`, therefore the patch is to fully qualify the type reference to the roslyn-analyzer's type.
+---
+ .../Core/MetaAnalyzers/Fixers/PreferIsKindFix.cs                | 2 +-
+ .../ApiDesignGuidelines/DoNotDirectlyAwaitATask.Fixer.cs        | 2 +-
+ 2 files changed, 2 insertions(+), 2 deletions(-)
+
+diff --git a/src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/Fixers/PreferIsKindFix.cs b/src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/Fixers/PreferIsKindFix.cs
+index f344b92fd..bc815d973 100644
+--- a/src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/Fixers/PreferIsKindFix.cs
++++ b/src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/Fixers/PreferIsKindFix.cs
+@@ -48,7 +48,7 @@ private async Task<Document> ConvertKindToIsKindAsync(Document document, TextSpa
+ 
+         protected abstract void FixDiagnostic(DocumentEditor editor, SyntaxNode nodeToFix);
+ 
+-        private sealed class CustomFixAllProvider : DocumentBasedFixAllProvider
++        private sealed class CustomFixAllProvider : Analyzer.Utilities.DocumentBasedFixAllProvider
```

Comment:
📝 This change is not fine. dotnet/roslyn-analyzers does not build against a version of dotnet/roslyn where this is a problem, and it is a configuration error to attempt to build dotnet/roslyn-analyzers against a different version of the reference assemblies.

Reply:
Just so I understand correctly the situation.  roslyn-analyzers currently runs with roslyn 4.0.0 (where this new type is exposed) in the 6.0 SDK but we prefer to compile it against an older version so that it can be installed with older SDK versions.  Is that correct?  If this is the case, we can certainly compile against older ref assembly to have parity between source-build and the repo build.  

To accommodate this, we will require changes to the roslyn-analyzer's versions.props file.  The reason being that sourcebuild's versions.props file contains the package versions it has produced.  The sourcebuild's versions.props entries override the roslyn-analyzer's versions.props file.  Since we want this to remain pinned, the name will need to be changed.

Reply:
> we prefer to compile it against an older version so that it can be installed with older SDK versions

Depending on the active development branch, this is sometimes the case. In other cases, we simply remain pinned for the purpose of avoiding churn from small changes to analyzer warnings or nullable signatures.

The main reason why we would want to continue taking action here is we want source build outputs to adhere as closely as technologically possible to the outputs produced by our primary builds. Deviations in dependency versions could have significant impact on interchangeability.

Reply:
> I believe an example here would clarify this.

Here is an example of the changes necessary to get roslyn-analyzers to build with a specific version during source-build rather than to utilize the source-built version 

https://github.com/MichaelSimons/roslyn-analyzers/commit/5759919a36072e091bf076578e113cd1bb3d2bc3

Reply:
@jaredpar it looks like source build was designed in a way that conflicts with our BuildBoss enforcement. Everything we currently have in Versions.props as `<PackageNameVersion>` needs to be changed to use `<PackageNameRefVersion>` instead, otherwise source build silently replaces the package version and produces binary outputs that no longer reflect our repository content.

Reply:
To clarify `<PackageNameVersion>` needs to be changed to something other than the Arcade package versions props naming convention.  You can choose the naming convention.  In my PR I used `<PackageNameRefVersion>` for illustration purposes.

Reply:
Can we fix the ordering of the versions.props used in source build to be included _before_ the repository file?

Reply:
>Can we fix the ordering of the versions.props used in source build to be included before the repository file? Seems like this would improve the results accuracy across the board.

No, this is not the behavior we want.  The norm is for repos to express their dependencies in the Versions.Details.xml and want the latest packages to flow.

Reply:
But that means source build isn't building the source. It's building an unknown variant of the source.

Reply:
Will ping @jaredpar on Teams

Reply:
> The norm is for repos to express their dependencies in the Versions.Details.xml and want the latest packages to flow.

That seems like the norm for teams that essentially track to the current runtime / sdk version. Essentially repositories that exist at the runtime layer of the .NET product. 

For teams that exist up at the SDK layer, particularly those that cross into the VS world, the norm is to pretty much never track to latest. The reason being that we're frequently supporting older versions of components and being inserted in servicing scenarios. Hence we're constantly referencing earlier versions of packages. 

Consider the roslyn compiler as an example here. Even though this ships in the SDK it never actually references the latest runtime. It is typically alwasy going to reference latest RTM TF - 1 (can elaborate on why if it's interesting but TLDR servicing is fun).

Reply:
A fair point.  The source-build team discussed this and we have an item on our backlog to make this space better so that we only flow the dependencies expressed in the versions.details.xml file.  This work is post 6.0 however.  We also discussed a few options that could be used in the interim to allow repos to import the source-build's versions.props before the repo's versions.props.  I will try this out and report back.

Reply:
@sharwell - I refactored the versions.props to explicitly import the source build versions.props at the beginning.  Please take a look.  There are some things we can do in Arcade / source-build to clean up this pattern as a follow-up if this approach is amenable.

Reply:
@sharwell, @jaredpar, @jmarolf - Please take a look at the refactored versions.props approach.